<a href="https://colab.research.google.com/github/PadmarajBhat/Machine-Learning/blob/master/BrainTumorClassification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Detection of 3 Brain Tumors (Meningioma, Glioma and Pituitary) in T1-weighted contrast enhanced images

### - Revisitng the Udacity Capstone Project in pursuit of better accuracy



# What is the problem statement?
  * predict the tumor class given a MRI image
  * OR predict the tumor class when both MRI and Tumor region is given !!!
      * tumor region is identified and put in input dataset by experts
          * can we have Image Segmentation problem ?


  * I think this is the order of problem from easy level to difficult level
    * Identify the tumor class from raw MRI image (here accuracy may be low)
    * Identify the tumor class from raw MRI image with tumor region identified info (here accuracy may be better)
    * Auto detect the tumor segment in a MRI image and classify the tumor (ideal application for a radiologist)

    Let us try all the 3 !!!

# Import Packages
* read the input MRI images (.mat) files through ***h5py***
* **bokeh** plot for the zoomed in analysis of a tumor and neighbors
* ***pandas*** for data analysis and preprocessing
* ***tensorflow*** for modelling and predicting

In [1]:
try:
  # %tensorflow_version only exists in Colab.
  %tensorflow_version 2.x
except Exception:
  pass

TensorFlow 2.x selected.


In [0]:
import os
import zipfile
import h5py
import numpy as np

import pandas as pd

from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler

from matplotlib import pyplot as plt
from bokeh.io import output_notebook, show
from bokeh.layouts import row
from bokeh.plotting import figure
output_notebook()

import imageio
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator

import random # for radom selection from a list
import datetime
import itertools




In [3]:
tf.__version__

'2.0.0-rc2'

In [0]:
# a Global variable
tumor_names = ["","Meningioma","Glioma","Pituitary"]

# Load Data

* **Iteration 1**:
    * Mount Google Drive
    * Unzip it in colab disk
    * load mat attributes to list of tuples ( with mri and tumor 5 point summary details)
    * create a panda dataframe for analysis

    ##### Issues Faced:
    1. loading to panda with image took half(6GB) of RAM
    * loading tumor along with mri image (as in mat file) crashed the colab
      * Solution: let us load image but save only 5 point summary for both mri image and tumor

    2. How do we scale/normalize the data?
      * would tumor region have 0 in it ?
        * only way to know is through the value present in the binary indicator == 1
            * implementation through 2 for loops takes forever !!!
              * need to implement throuhg np.where...a[a == 1] !!!
            

    3. Some images are less than 512
        * pad the difference with 0s.
        
    4. Should tumor image be scaled between 0 -1? For now, brightness values are relative to that of the whole image to which it belongs to.

    5. Epoch run failed due to no data generated by the custom generator.
      * Going to try the ImageGenerator from the TF.


* **Iteration 2** :
  * ImageGenerator worked fine but the np to image conversion had used dtype of np.uint8 to avoid warning during saving to image. However, that lead to corrupt image and hence loss was more and accuracy was less.
  * Generator built for the iteration is correct ?
      * are *flips* is not damaging data

* **Iteration 3** :
  * validate the image augmentation in the ImageGenerator
  * initial load the data was fine. zip based train and test failed

* **Iteration 4**:
  * open all mat(hdf5) files and load 5 point summary of mri and tumor to a panda df
  * save all the numpy mri image array to training_data directory
  * split the df to training : testing = 80 :20
  * move the 20% of the testing to testing_data
  * out of the 80% training data, move the 10% for the validation during training
      * download
        * 5
        * b*.zip
        * *.txt
        * mat
          * *.mat

      * training_data
        * 1
          * *.npy
        * 2
          *  *.npy
        * 3
          * *.npy
      * validation_data
        * 1
          * *.npy
        * 2
          *  *.npy
        * 3
          * *.npy
      * testing_data
        * 1
          * *.npy
        * 2
          *  *.npy
        * 3
          * *.npy


  

from google.colab import drive
drive.mount('/content/drive')

##### Google Drive File Check

In [0]:
!rm -rf training_data validation_data testing_data download

In [21]:
def loadMatFiles(dir="training_data/"):
  if not os.path.isfile("download/mat/1.mat"):
    cmds = [
            "rm -rf download data training_data validation_data testing_data"
            ,"wget https://ndownloader.figshare.com/articles/1512427/versions/5 -P download"
            ,"unzip download/5 -d download"
            ,"unzip -q download/brainTumorDataPublic_1-766.zip -d download/mat"
            ,"unzip -q download/brainTumorDataPublic_1533-2298.zip -d download/mat"
            ,"unzip -q download/brainTumorDataPublic_767-1532.zip   -d download/mat"
            ,"unzip -q download/brainTumorDataPublic_2299-3064.zip  -d download/mat"]

    for c in cmds:
      os.system(c)
    time.sleep(5)
  else:
    print("mat files are loaded into download/mat directory")

loadMatFiles()
print("Total Data : ", len(os.listdir("download/mat")))

Total Data :  3064


In [23]:
if not os.path.isdir("training_data/"):

  if not os.path.isdir("download/mat") or not len(os.listdir("download/mat")) :
    loadMatFiles()

  os.system("mkdir training_data validation_data testing_data")

  files = os.listdir("download/mat").copy()
  training_files = random.sample(files, k=round(len(files) *.7))

  for file in training_files:
    try:
      os.rename("download/mat/"+file, "training_data/"+file)
    except:
      print("Skipping :", file)
  time.sleep(5)

print("Training Data:", len(os.listdir("training_data/")))

Training Data: 2145


In [25]:
files = os.listdir("download/mat").copy()
validation_files = random.sample(files, k=round(len(files) *.1))

for file in validation_files:
  try:
    os.rename("download/mat/"+file, "validation_data/"+file)
  except:
    print("Skipping :", file)
time.sleep(5)
print("Validation Data:", len(os.listdir("validation_data/")))

Validation Data: 175


In [26]:
for file in os.listdir("download/mat/"):
      if not (file in os.listdir("training_data/") or file in os.listdir("validation_data")):
        try:
          os.rename("download/mat/"+file, "testing_data/"+file)
        except:
          print("Skipping :", file)
time.sleep(5)
print("Testing Data:", len(os.listdir("testing_data/")))

Testing Data: 744


In [0]:
def returnImageLabel(loc, file_list):
  image_list=[]
  label_list=[]
  label_transform = [[],np.array([0,0,1]),np.array([0,1,0]),np.array([1,0,0])]
  for file_name in file_list:
    
    with h5py.File(loc+file_name,'r') as f:
          image_array = np.array(f['cjdata']['image'],dtype=np.float64)
          image_array = image_array/image_array.max()
          label = np.array(f['cjdata']['label'], dtype=np.int)[0][0]
          if image_array.shape[0] != 512:
            image_array = np.pad(image_array,(512 - image_array.shape[0])//2,'constant',constant_values=0)
          
          #image_list.append((image_array,  label_transform[label]))
          image_list.append(image_array)
          label_list.append(label_transform[label])
  return np.array(image_list), np.array(label_list)
  #df = pd.DataFrame(image_list,columns=["image_array","label"])
  #return df

def myTrainGenerator(batch_size):
  files =os.listdir("training_data/npz/")
  for i in itertools.cycle(files):
    data = np.load("training_data/npz/"+i)
    yield data['x'].reshape((data['x'].shape[0],512*512)), data['y']

def myValidateGenerator(batch_size):
  files = os.listdir("validation_data/npz/") 
  for i in itertools.cycle(files):
    data = np.load("validation_data/npz/"+i)
    yield data['x'].reshape((data['x'].shape[0],512*512)), data['y']

def myTestGenerator(batch_size):
  files =  os.listdir("testing_data/npz/")
  for i in files:
    data = np.load("testing_data/npz/"+i)
    yield data['x'].reshape((data['x'].shape[0],512*512)), data['y']



In [28]:
def npSaver(loc):
  
  files = os.listdir(loc).copy()
  os.system("mkdir "+loc+"npz/")
  for i in range(0,len(files),32):
    x,y = returnImageLabel(loc, files[i:i+32])
    print(loc+"npz/"+str(i)+".npz")
    np.savez(loc+"/npz/"+str(i), x=x, y=y)
    #time.sleep(1)
    #np.save(loc+str(i),np.array(returnImageLabel(loc, files[i:i+32])))

if not os.path.isfile("training_data/npz/0.npz"):
  npSaver("training_data/")
print("Training Batch Files :", len(os.listdir("training_data/npz")))

if not os.path.isfile("validation_data/npz/0.npz"):
  npSaver("validation_data/")
print("Validation Batch Files :", len(os.listdir("validation_data/npz")))

if not os.path.isfile("testing_data/npz/0.npz"):
  npSaver("testing_data/")
print("Testing Batch Files :", len(os.listdir("testing_data/npz")))


training_data/npz/0.npz
training_data/npz/32.npz
training_data/npz/64.npz
training_data/npz/96.npz
training_data/npz/128.npz
training_data/npz/160.npz
training_data/npz/192.npz
training_data/npz/224.npz
training_data/npz/256.npz
training_data/npz/288.npz
training_data/npz/320.npz
training_data/npz/352.npz
training_data/npz/384.npz
training_data/npz/416.npz
training_data/npz/448.npz
training_data/npz/480.npz
training_data/npz/512.npz
training_data/npz/544.npz
training_data/npz/576.npz
training_data/npz/608.npz
training_data/npz/640.npz
training_data/npz/672.npz
training_data/npz/704.npz
training_data/npz/736.npz
training_data/npz/768.npz
training_data/npz/800.npz
training_data/npz/832.npz
training_data/npz/864.npz
training_data/npz/896.npz
training_data/npz/928.npz
training_data/npz/960.npz
training_data/npz/992.npz
training_data/npz/1024.npz
training_data/npz/1056.npz
training_data/npz/1088.npz
training_data/npz/1120.npz
training_data/npz/1152.npz
training_data/npz/1184.npz
training_da

[os.system("rm -rf "+"training_data/"+f) for f in os.listdir("training_data/") if ".np" in f or ".csv" in f]
[os.system("rm -rf "+"validation_data/"+f) for f in os.listdir("validation_data/") if ".np" in f or ".csv" in f]
[os.system("rm -rf "+"testing_data/"+f) for f in os.listdir("testing_data/") if ".np" in f or ".csv" in f]
!rm -rf "training_data/npz validation_data/npz testing_data/npz"

### Load Image Array

def retrieveImage(file_name):
  f = h5py.File(file_name,'r')
  mri_image = np.array(f['cjdata']['image'],dtype=np.float64)
  if mri_image.shape[0] < 512:
      print("Shape of the image : ", mri_image.shape)
      mri_image = np.pad(mri_image,(512 - mri_image.shape[0])//2,'constant',constant_values=0)
  return mri_image/mri_image.max()

### Load Tumor Array

def retrieveTumorImage(file_name):
  f = h5py.File(file_name,'r')
  mri_image = np.array(f['cjdata']['tumorMask'],dtype=np.float128)
  if mri_image.shape[0] < 512:
      print("Shape of the image : ", mri_image.shape)
      mri_image = np.pad(mri_image,(512 - mri_image.shape[0])//2,'constant',constant_values=0)
  return mri_image/mri_image.max()

### Create Directories for ImageGenerator

!mkdir "training_data/images/"
!mkdir "training_data/images/1"
!mkdir "training_data/images/2"
!mkdir "training_data/images/3"

!mkdir "testing_data/images/"
!mkdir "testing_data/images/1"
!mkdir "testing_data/images/2"
!mkdir "testing_data/images/3"

### Load Image and Tumor Statistics to Panda

*   Panda df would have 5 point summary of both mri and tumor
*   data directory will have label wise subdirectories for ImageGenerator



def return_imageInfo_from_mat_file(dir,file_name):
    f = h5py.File(dir+file_name,'r')

    mri_image = np.array(f['cjdata']['image'],dtype=np.float128)
    #scaler = MinMaxScaler(feature_range=(1,2))
    #mri_image = scaler.fit(mri_image)
    mri_image = mri_image/mri_image.max()

    if mri_image.shape[0] < 512:
      print("Shape of the image : ", mri_image.shape)
      mri_image = np.pad(mri_image,(512 - mri_image.shape[0])//2,'constant',constant_values=0)
    
    temp_mri_image = np.copy(mri_image)
    temp_mri_image[temp_mri_image == 0 ] = 2

    mri_quartiles = np.percentile(mri_image[mri_image > 0], [25, 50, 75,80,85,90,95,96,97,98,99])

    tumor_image = np.array(f['cjdata']['tumorMask'], dtype=np.float128)
    if tumor_image.shape[0] < 512:
      print("Shape of the tumor image : ", tumor_image.shape)
      tumor_image = np.pad(tumor_image,(512 - tumor_image.shape[0])//2,'constant',constant_values=0)
    
    tumor_image = temp_mri_image * tumor_image
    tumor_image = tumor_image[tumor_image > 0]
    tumor_image[tumor_image == 2] = 0

    '''tumor_array =[]
    for i in range(0,512):
      for j in range(0,512):
        if tumor_image[i][j]:
          tumor_array.append(mri_image[i][j])

    tumor_image = np.array(tumor_array, dtype=np.float)'''

    tumor_quartiles = np.percentile(tumor_image, [25, 50, 75,80,85,90,95,96,97,98,99])

    label=np.array(f['cjdata']['label'], dtype=np.int)[0][0]
    imageio.imwrite(dir+"images/"+str(label)+"/"+file_name.split(".")[0]+'.jpg', np.array(f['cjdata']['image'],dtype=np.int16))

    return np.array(f['cjdata']['PID'],dtype=np.int)[0][0] \
            ,mri_image.min() \
            ,mri_image.max() \
            ,mri_quartiles[0] \
            ,mri_quartiles[1] \
            ,mri_quartiles[2] \
            ,mri_quartiles[3] \
            ,mri_quartiles[4] \
            ,mri_quartiles[5] \
            ,mri_quartiles[6] \
            ,mri_quartiles[7] \
            ,mri_quartiles[8] \
            ,mri_quartiles[9] \
            ,mri_quartiles[10] \
            ,tumor_image.min() \
            ,tumor_image.max() \
            ,tumor_quartiles[0] \
            ,tumor_quartiles[1] \
            ,tumor_quartiles[2] \
            ,tumor_quartiles[3] \
            ,tumor_quartiles[4] \
            ,tumor_quartiles[5] \
            ,tumor_quartiles[6] \
            ,tumor_quartiles[7] \
            ,tumor_quartiles[8] \
            ,tumor_quartiles[9] \
            ,tumor_quartiles[10] \
            ,tumor_image.shape \
            ,file_name\
            ,np.array(f['cjdata']['label'], dtype=np.int)[0][0] 

def loadDf(dir="training_data/"):
  patients_details = []
  '''for root, dirs, files in os.walk("/content/drive/My Drive/1512427/", topdown = False):
    for f in files:
      if ".zip" in f:
          file = zipfile.ZipFile(root+f, "r")
          for name in file.namelist():
            file.extract(name,".")
            patients_details.append(return_imageInfo_from_mat_file(name))
          #break
      #break  '''   
  
  for f in getFileNames(dir):
    if ".mat" in f:
      patients_details.append(return_imageInfo_from_mat_file(dir,f))
  mri_col_names = ["mri_min","mri_max","mri_1q","mri_median", "mri_3q","mri_80","mri_85","mri_90","mri_95","mri_96","mri_97","mri_98","mri_99"]
  tumor_col_names = ["t_min","t_max","t_1q","t_median","t_3q","t_80","t_85","t_90","t_95","t_96","t_97","t_98","t_99","tumor_size"]
  col_names = ["pid"] + mri_col_names + tumor_col_names+ ["file_name","label"]
  return pd.DataFrame(patients_details,columns=col_names)


### Load Training Data

df = loadDf()
df["square_shape"] = df.tumor_size.apply(lambda x: np.sqrt(x[0]))
df.sample(20)

### Load Testing Data

df_test = loadDf("testing_data/")
df_test["square_shape"] = df_test.tumor_size.apply(lambda x: np.sqrt(x[0]))
df_test.sample(20)

### Test the loaded data (both Training and Testing Data)

def displayNpImages(dir="training_data/"):
  for i in range(1,4):
    print("2 samples for ",tumor_names[i])
    for j in random.choices( os.listdir(dir+str(i)),k=2):
      plt.imshow(plt.imread(dir+str(i)+"/"+str(j)))
      plt.show()

#displayNpImages()

#### Visual Testing Testing Data

In [0]:
#displayNpImages("testing_data/")

In [0]:
#displayNpImages("validation_data/")

def analyzeZipDir(df,start,end):
  df["file_num"] = df.file_name.apply(lambda x: x.split(".")[0])
  df["file_num"] = df.file_num.astype(np.int)
  
  return df[df.file_num.isin(list(range(start,end)))]
analyzeZipDir(df.copy(),1,766).groupby("label").agg("count").reset_index()[["label","pid"]]

analyzeZipDir(df.copy(),1533,2298).groupby("label").agg("count").reset_index()[["label","pid"]]

analyzeZipDir(df_test.copy(),2299,3064).groupby("label").agg("count").reset_index()[["label","pid"]]

analyzeZipDir(df.copy(),767,1532).groupby("label").agg("count").reset_index()[["label","pid"]]

### Create Test directory for validation through generators

!rm -rf "test"
!mkdir "test"
!mkdir "test/1"
!mkdir "test/2"
!mkdir "test/3"

!ls -l /content/data/2/3046.jpg
!ls -l /content/test

!ls -l /content/data/2/2404.jpg

### Move the test set to test directory from data directory

import shutil
import random

for root, dirs, files in os.walk("/content/data", topdown = False):
  
    
    if len(files) > 0:
      print(root, dirs, files)

      #indices = np.random.randint(0,len(files),size=round(len(files)*.2))
      rand_files = random.choices(files,k=round(len(files)*.2))
      
      for f in rand_files:
        #print(f)
        try:
          shutil.move(root+"/"+f, "/content/test/"+root.split("/")[-1]+"/"+f)
        except :
          print("Ignoring : ",f)

#list(os.walk("/content/data")) /content/test /content/training_data

### ImageGenerators

### Load tf.data
* inspired from: https://stackoverflow.com/questions/48309631/tensorflow-tf-data-dataset-reading-large-hdf5-files
* extract all files
* save file names in a tensor
* write a generator function to read a file and return numpy mri image array and its label
* using tf.dataset.interleave function, read the file concurrently
* apply batch and shuffle
* feed it to model


##### Download the MRI image zip

def mygenerator(file_name, dir="training_data/"):
  #print(file_name)
  f = h5py.File(dir+file_name,'r')
  #print(f['cjdata']['image'].dtype)
  mri_image = np.array(f['cjdata']['image'],dtype=np.int16)
  if mri_image.shape[0] < 512:
      #print("Shape of the image : ", mri_image.shape)
      mri_image = np.pad(mri_image,(512 - mri_image.shape[0])//2,'constant',constant_values=0)
  return (mri_image, np.array(f['cjdata']['label'], dtype=np.int)[0][0])


df_temp = pd.DataFrame([mygenerator(f) for f in getFileNames()],columns=["image","label"])

'''ds = tf.data.Dataset.from_tensor_slices([  tf.data.Dataset.from_generator(
        mygenerator(filename), 
        (tf.uint8,tf.int8), 
        (tf.TensorShape([]), tf.TensorShape([None]))) for filename in ds])'''
df_temp.head()

imageio.imwrite("test.jpg",df_temp.iloc[0][0])
plt.imshow(imageio.imread("test.jpg"),cmap='bone')
plt.show()
plt.imshow(plt.imread("test.jpg"))
plt.show()
plt.imshow(df_temp.iloc[0][0])

plt.imshow(np.array(df_temp.iloc[0][0],dtype=np.uint8))

class ToTensor:
  def __init__(self):
    print("Obj Created")
  
  def unzipData(self,source_dirname, dest_dir="data/"):
    try:
      shutil.rmtree(dest_dir)
    except:
      pass
    os.mkdir(dest_dir)
    print("Exploring ",source_dirname, "directory")
    for root, dirs, files in os.walk(source_dirname, topdown = False):
      for f in files:
        print("Found file ",f)
        if ".zip" in f:
          print("Unzipping ", f)
          with zipfile.ZipFile(f) as zf:
            zf.extractall(dest_dir)
    print("Files Loaded !!!")
    return self

  def getFileName(self,dir_name="data/"):
    return os.listdir(dir_name)

toTensor = ToTensor()
toTensor.unzipData("/content/drive/My Drive/1512427/").getFileName()

# Analysis


## Statistical Analysis
* Number of patients in the dataset
* Patient wise distribution of tumor classes
* Comparison of below attributes for the 3 tumor classes
  * 1st quantile of MRI image
  * Median of the MRI image
  * 3rd quantile of the MRI image
  * min value distribution of the Tumor
  * max value distribution of the Tumor
  * 1st quantile of the Tumor
  * median of the Tumor
  * 3rd quantile of the Tumor
    * Analysis: 
      * All tumors have darkest area which may indicate the tumor itself
      * All tumors have uniform distribution of brightness (apart from the dark area)
      * MRI images have darker area outside the skull (non scan area)
          * will this influence the model ?
          * should the color of the tumor and the non important area of the MRI scan be different ?
          
* 256x256 image size distribution (any bias in there ?)

##### Issues Faced:
* Bokeh plots are interactive but they consume a lot of space(>100mb) in the notebook 
  * markdown for now, when interested can be seen by enabling it as code cell

df.pid.unique()

There are only 5 patients info present !!!!

df.groupby("pid").agg("count").reset_index()[['pid','mri_min']]

df.groupby(["pid","label"]).agg("count").reset_index()[['pid','label','mri_min']]

df.groupby("label").agg("count").reset_index()[['label','pid']]

def plotStatistics(df, tumor_name):
  df = df[["mri_1q","mri_median","mri_3q","t_min","t_1q","t_median","t_3q","t_max"]]
  df=(df-df.min())/(df.max()-df.min())
  fig, ax = plt.subplots(1, 8,sharex=True,sharey=True,tight_layout=True)
  fig.set_figheight(4)
  fig.set_figwidth(13)
  
  fig.suptitle(tumor_name+" Tumor")
  #plt.subplot(1,8,1)
  ax[0].hist(df.mri_1q.tolist())
  ax[0].set_title("mri_1q")
  #plt.subplot(1,8,2)
  ax[1].hist(df.mri_median.tolist())
  ax[1].set_title("mri_median")
  #plt.subplot(1,8,3)
  ax[2].hist(df.mri_3q.tolist())
  ax[2].set_title("mri_3q")
  #plt.subplot(1,8,4)
  ax[3].hist(df.t_min.tolist())
  ax[3].set_title("t_min")
  #plt.subplot(1,8,5)
  ax[4].hist(df.t_1q.tolist())
  ax[4].set_title("t_1q")
  #plt.subplot(1,8,6)
  ax[5].hist(df.t_median.tolist())
  ax[5].set_title("t_median")
  #plt.subplot(1,8,7)
  ax[6].hist(df.t_3q.tolist())
  ax[6].set_title("t_3q")
  #plt.subplot(1,8,8)
  ax[7].hist(df.t_max.tolist())
  ax[7].set_title("t_max")
  plt.show()

plotStatistics(df[df.label ==1], tumor_names[1])
plotStatistics(df[df.label ==2], tumor_names[2])
plotStatistics(df[df.label ==3], tumor_names[3])
plotStatistics(df, "All")

## Can we reduce the image size ?
* Why?
  * faster model building
  * lower convolution experiment iterations
  * lower ram usage and hence higher batch size



### Approach 1 : Can we segregate the skull ?
  * removing the unwanted area 
    * percentile approach: identify the percentile and see if any of the tumor percentile is always less that the MRI percentile. 
        i.e. to prove mri_99 > t_99. This failed as indicated below
    * brightness based skull identification:
      * nearest neighbor ???

df[df.mri_99 < df.t_99]

plt.imshow(retrieveImage("120.mat"))
plt.show()

### Approach 2: PCA
* Note that if we do the PCA transformation, we 2D image will be reduced to 1D. Therefore, we can not use it for the Convolution approach.
  * we can note here that just by 2 features (components) and the trained PCA model, we are able to recreate the image with not much difference. [See the last 3 images] This showcases the PCA strength.

* Withhelding the PCA approach as we are going to pursuit the Convolution 


import numpy as np
from sklearn.decomposition import PCA
pca = PCA(n_components=5,whiten=True)
image=[]
image.append(retrieveImage("120.mat").reshape(-1))
image.append(retrieveImage("1.mat").reshape(-1))
image.append(retrieveImage("2.mat").reshape(-1))
image.append(retrieveImage("3.mat").reshape(-1))
image.append(retrieveImage("4.mat").reshape(-1))
image.append(retrieveImage("5.mat").reshape(-1))
image.append(retrieveImage("6.mat").reshape(-1))
image.append(retrieveImage("7.mat").reshape(-1))

#print(image.shape)

pca.fit(image)
plt.imshow(pca.mean_.reshape((512,512)),
           cmap=plt.cm.bone)
plt.show()

print(pca.noise_variance_)
print(image[0].reshape((1,-1)).shape)
pca.transform(image[0].reshape((1,-1)))

#plt.imshow(pca.transform(image[1].reshape(1,-1)).reshape((512,512)),cmap=plt.cm.bone)


components = pca.transform(image[0].reshape(1,-1))
projected = pca.inverse_transform(components)
plt.imshow(projected.reshape((512,512)))
plt.show()
plt.imshow(image[0].reshape((512,512)))
plt.show()
plt.imshow(retrieveImage("120.mat"))
plt.show()
print("doe it match :", projected.reshape((512,512)) == retrieveImage("120.mat"))

plt.imshow(image[0].reshape((256,256)))
plt.show()

## Visual Analysis

### Smallest Tumor Sample

plt.imshow(retrieveImage(list(df[df.tumor_size == df.tumor_size.min()]['file_name'])[0]));
plt.imshow(retrieveTumorImage(list(df[df.tumor_size == df.tumor_size.min()]['file_name'])[0]),alpha=0.5);
plt.show()


### Biggest Tumor in the Dataset

plt.imshow(retrieveImage(list(df[df.tumor_size == df.tumor_size.max()]['file_name'])[0]));
plt.imshow(retrieveTumorImage(list(df[df.tumor_size == df.tumor_size.max()]['file_name'])[0]),alpha=0.5);
plt.show()


### Numpy Resize failed

plt.imshow(np.resize(retrieveImage(list(df[df.tumor_size == df.tumor_size.max()]['file_name'])[0]),(256,256)));
plt.show()

### Bokeh Plot

def bokehPlot(file_name, tumor_label):
  tumor_names = ["","Meningioma","Glioma","Pituitary"]
  im = retrieveImage(file_name)
  s1 = figure(width=512, plot_height=512, title=tumor_names[tumor_label]+" MRI Image")
  s1.image([im],x=[0],y=[0],dw=[512],dh=[512])

  im2 = retrieveTumorImage(file_name)

  s2 = figure(width=500, plot_height=500, title=tumor_names[tumor_label]+" MRI Image with Tumor Highlighted")
  s2.image([im2],x=[0],y=[0],dw=[512],dh=[512])
  s2.image([im],x=[0],y=[0],dw=[512],dh=[512],global_alpha=0.5)

  show(row(s1,s2))

bokehPlot(list(df[df.tumor_size == df.tumor_size.max()]['file_name'])[0], list(df[df.tumor_size == df.tumor_size.max()]['label'])[0])

#### Meningioma Plots

for fname in list(df[df.label == 1].sample(3)["file_name"]):
  bokehPlot(fname,1)

#### Glioma Plots

for fname in list(df[df.label == 2].sample(3)["file_name"]):
  bokehPlot(fname,2)

#### Pituitary Plots

for fname in list(df[df.label == 3].sample(3)["file_name"]):
  bokehPlot(fname,3)

# Preprocessing


Preprocessing ideas:

1.  Dataset has tumor region indicator which would allow us to get the average brightness of the area.

2. It is said that brightest region is skull and skull is not important for the tumor detection. It is only brain position determines the tumor class. If we remove skull remaining image is brain ?

3. if we start with a window of image which would maximize the presence of tumor and expand to include some brain region around the tumor then i guess it is the best data for training(and predicting). Because tumor position in brain is THE factor that decides the tumor class.

4. what is the optimum batch size for training?

5. what is the overall Image augumented training dataset size ?



## Train & Test split



def getSplit(df):
  df_test=df.sample(frac=.2)
  df = df.drop(df_test.index)
  return df, df_test

df_orig = df.copy()
df,df_test = getSplit(df)

## Batch Creation

df.groupby("label").agg("count").reset_index()

def returnBatchIndices(df,batch_size):
  label_1 = df[df.label == 1].index.tolist()
  label_2 = df[df.label == 2].index.tolist()
  label_3 = df[df.label == 3].index.tolist()

  label_list = []
  #print(len(label_1), len(label_2),len(label_3),list(range(0,max(len(label_1),len(label_2),len(label_3)),batch_size)))
  for i in range(0,max(len(label_1),len(label_2),len(label_3)),batch_size):
    label_list.append(label_1[i:i+batch_size] + label_2[i:i+batch_size] + label_3[i:i+batch_size])
  return label_list

#yieldbatch(df,5)
for batch in returnBatchIndices(df,5):
  print(batch)
  break

print("Total Number of Batches: ", len(returnBatchIndices(df,5)))

### For Convolution

def returnABatch(df,batch_size):
  #returns a balanced label mri images
  index_list = returnBatchIndices(df,batch_size)
  #print("index list",len(list(index_list)))
  df2 = pd.get_dummies(df['label'], prefix = 'label')
  df = pd.concat([df,df2],axis=1)
  for j in index_list:
    batch_images=[]
    batch_labels=[]
    #print("j",j)
    for i in j:
      #print("i",i)
      label_list=[]
      image = retrieveImage(list(df[df.index == i]['file_name'])[0])
      transformed_image = image.reshape((512,512,1))
      batch_images.append(transformed_image)
      label_list.append(df[df.index == i]['label_1'].tolist()[0])
      label_list.append(df[df.index == i]['label_2'].tolist()[0])
      label_list.append(df[df.index == i]['label_3'].tolist()[0])
      batch_labels.append(label_list)
      #print("Batches :",len(batch_images),len(batch_labels))

    #from keras.utils import to_categorical
    #batch_labels = to_categorical(batch_labels)
    yield np.array(batch_images), np.array(batch_labels)

for i in returnABatch(df.reset_index(),2)  :
  if i[1].shape[0] < 6:
    print(len(i),len(i[1]))
    print(i[0].shape)
    print(i[1])
    break

### For Logistic Regression

def returnABatch1d(df,batch_size):
  #returns a balanced label mri images
  index_list = returnBatchIndices(df,batch_size)
  #print("index list",len(list(index_list)))
  df2 = pd.get_dummies(df['label'], prefix = 'label')
  df = pd.concat([df,df2],axis=1)
  for j in index_list:
    batch_images=[]
    batch_labels=[]
    #print("j",j)
    for i in j:
      #print("i",i)
      label_list=[]
      image = retrieveImage(list(df[df.index == i]['file_name'])[0])
      transformed_image = image.reshape(512*512)
      batch_images.append(transformed_image)
      label_list.append(df[df.index == i]['label_1'].tolist()[0])
      label_list.append(df[df.index == i]['label_2'].tolist()[0])
      label_list.append(df[df.index == i]['label_3'].tolist()[0])
      batch_labels.append(label_list)
      #print("Batches :",len(batch_images),len(batch_labels))

    #from keras.utils import to_categorical
    #batch_labels = to_categorical(batch_labels)
    yield np.array(batch_images), np.array(batch_labels)

for i in returnABatch1d(df.reset_index(),2)  :
  if i[1].shape[0] < 6:
    print(len(i),len(i[1]))
    print(i[0].shape)
    print(i[1])
    break

#Model Building


## CNN Approach using Tensorflow keras

In [9]:
model = tf.keras.models.Sequential([
            tf.keras.layers.Conv2D(64, (3,3), activation="relu", input_shape=(512,512,3))
            ,tf.keras.layers.MaxPooling2D(2,2)
            ,tf.keras.layers.Conv2D(64, (3,3), activation="relu")
            ,tf.keras.layers.MaxPooling2D(2,2)
            ,tf.keras.layers.Conv2D(64, (3,3), activation="relu")
            ,tf.keras.layers.MaxPooling2D(2,2)
            ,tf.keras.layers.Conv2D(64, (3,3), activation="relu")
            ,tf.keras.layers.MaxPooling2D(2,2)
            ,tf.keras.layers.Conv2D(64, (3,3), activation="relu")
            ,tf.keras.layers.MaxPooling2D(2,2)
            #,tf.keras.layers.Conv2D(64, (3,3), activation="relu")
            #,tf.keras.layers.MaxPooling2D(2,2)
            ,tf.keras.layers.Flatten()
            ,tf.keras.layers.Dropout(0.5)
            ,tf.keras.layers.Dense(512, activation="relu")
            ,tf.keras.layers.Dense(3,activation="softmax")            
])

model.compile(loss="categorical_crossentropy"
              ,optimizer= "adam"
              ,metrics=["accuracy","Precision","Recall"])
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 510, 510, 64)      1792      
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 255, 255, 64)      0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 253, 253, 64)      36928     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 126, 126, 64)      0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 124, 124, 64)      36928     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 62, 62, 64)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 60, 60, 64)        3

img = plt.imread("/content/data/1/1.jpg")
plt.imshow(img,cmap="bone")

target = df_temp.pop('label')
dataset = tf.data.Dataset.from_tensor_slices((df_temp.values, target.values))

In [10]:
history=model.fit_generator(train_generator
                  #,steps_per_epoch=286
                  , epochs=2
                  ,use_multiprocessing=True
                  ,validation_data=validate_generator
                  ,workers=8
                  )

NameError: ignored

In [0]:
model.save('./final_model.h5', include_optimizer=True)
from tensorflow.keras.models import load_model
model2 = load_model('./final_model.h5')
history2=model2.fit_generator(train_generator
                  #,steps_per_epoch=286
                  , epochs=2
                  ,use_multiprocessing=True
                  ,validation_data=validate_generator
                  ,workers=8
                  )

result2 = model2.evaluate(test_generator)
print(result2)

In [0]:
result = model.evaluate(test_generator)
print(result)

In [0]:
history=model.fit_generator(validate_generator
                  #,steps_per_epoch=286
                  , epochs=2
                  ,use_multiprocessing=True
                  ,validation_data=test_generator
                  ,workers=8
                  )
result = model.evaluate(test_generator)
print(result)


### Pretrained model

In [0]:
def preTrain():
  import os

  from tensorflow.keras import layers
  from tensorflow.keras import Model
  !wget --no-check-certificate \
      https://storage.googleapis.com/mledu-datasets/inception_v3_weights_tf_dim_ordering_tf_kernels_notop.h5 \
      -O /tmp/inception_v3_weights_tf_dim_ordering_tf_kernels_notop.h5
  from tensorflow.keras.applications.inception_v3 import InceptionV3

  local_weights_file = '/tmp/inception_v3_weights_tf_dim_ordering_tf_kernels_notop.h5'

  pre_trained_model = InceptionV3(input_shape = (512,512,3), 
                                  include_top = False, 
                                  weights = None)

  pre_trained_model.load_weights(local_weights_file)

  for layer in pre_trained_model.layers:
    layer.trainable = False
    
  #print(pre_trained_model.summary())

  last_layer = pre_trained_model.get_layer('mixed7')
  print('last layer output shape: ', last_layer.output_shape)
  last_output = last_layer.output

  from tensorflow.keras.optimizers import RMSprop
  x = layers.Conv2D(64, (3,3), activation="relu")(last_output)
  x = layers.MaxPooling2D(2,2)(x)
  x = layers.Conv2D(64, (3,3), activation="relu")(x)
  x = layers.MaxPooling2D(2,2)(x)
  # Flatten the output layer to 1 dimension
  #x = layers.Flatten()(last_output)
  x = layers.Flatten()(x)
  # Add a fully connected layer with 1,024 hidden units and ReLU activation
  x = layers.Dense(512, activation='relu')(x)
  # Add a dropout rate of 0.2
  x = layers.Dropout(0.5)(x)                  
  # Add a final sigmoid layer for classification
  x = layers.Dense  (3, activation='sigmoid')(x)           

  model = Model( pre_trained_model.input, x) 
  return model
imageNet_model = preTrain()

In [0]:
imageNet_model.summary()

In [0]:
imageNet_model.compile(loss="categorical_crossentropy"
              ,optimizer= "adam"
              ,metrics=["accuracy","Recall","Precision"])

In [0]:
train_generator, validate_generator, test_generator = getGenerators(32,(512,512))
history=imageNet_model.fit_generator(train_generator
                  #,steps_per_epoch=286
                  , epochs=3
                  ,use_multiprocessing=True
                  ,validation_data=validate_generator
                  #,workers=2
                  )

In [0]:
result = imageNet_model.evaluate(test_generator)
print(result)

## Logistic Regression using Tensorflow Keras

In [0]:
model = tf.keras.models.Sequential([
             tf.keras.layers.Flatten(input_shape=(512, 512,3))
            ,tf.keras.layers.Dense(3, activation="softmax")            
])

model.compile(loss="categorical_crossentropy"
              ,optimizer= "adam"
              ,metrics=["accuracy","Recall","Precision"])
model.summary()

In [0]:
history=model.fit_generator(train_generator
                  #,steps_per_epoch=286
                  , epochs=5
                  ,use_multiprocessing=True
                  ,validation_data=validate_generator
                  ,workers=2
                  )

In [0]:
result = model.evaluate(test_generator)
print(result)

batch_size = 64
steps_per_epoch = round(df.groupby("label").agg("count").reset_index()['pid'].max()/batch_size)

print("Total Training Dataset : ", df.shape[0])
print("Batch Size : ", batch_size)
print("Steps per epoch : ", steps_per_epoch)
print("Test Datasize shape : ", df_test.shape[0])

history=model.fit_generator(returnABatch1d(df,batch_size)
                  ,steps_per_epoch=steps_per_epoch
                  , epochs=5
                  )

In [0]:
# summarize history for accuracy
plt.plot(history.history['acc'])
plt.plot(history.history['val_acc'])
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'test'], loc='upper left')
plt.show()
# summarize history for loss
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'test'], loc='upper left')
plt.show()

# Model Testing
* is the model bias for 256 size images ?
* is there any imbalance in 256 size images ?
* converting 512x512 to 256x256 size would definitely speed up the process but would it impact the accuracy ?
* is the model has better accuracy for any type of tumor class? (as we have imbalanced set ?

# Observations / Lesson Learnt:

* Iteration 1:
  * CNN of 512x512 took half an hour even on TPU
  * more and more convolution layer decreases the neurons required for training( duh!!!) and hence the batch size can be increased.
  * testing result was 49%. Not Acceptable.

* Iteration 2:
  * tried PCA but realized transformation is not suited for convolution
  * tried to use imagegenerator by converting the numpy to image (jpg) file. But found that numpy to image was not successfull.

* Iteration 3:
  * experimented with ImageGenerator for training, validation and testing
  * time is still a worry for CNN approach; it is 20 mins
  * though it is fast for the logistic regression the epochs are not helping (epoch = 5)


* Iteration 4:
  * experimented with different CNN layers ( basically added one more layer to make shrink the image size to 6x6 but 14x14 was quicker and gave better result)
  * average time per epoch is 20 min and 5 epochs gave 89% test accuracy with CNN approach
  * deviation in the epoch accuracy indicates that few more epoch can reach upto 90 - 93% which way ahead of last finding during capstone submission (54%).[validation accuracy has also increased so far and hence predicting the accuracy to go beyond 90%]
  * imagenet transfer learing takes 1 hour to complete one epoch and for one epoch it gives accuracy around 45%
  * 3 epochs on imagenet with 512x512x3 image with 2 convolution at the end with dense network post that gave 90% training accuracy, 65% validation and testing result. Clearly, there is overfit.
  * shrinking the size by 256x256 or 150x150 (as per imagenet assumption on input), drastically reduces the epoch time upto 5 min but the accuracy do not improve even for 5 epochs. Loss decreases but there is no change in the accuracy or accuracy decreases.
  * till now, it leaves with no additional lesson learnt apart from few experimental observations. i.e. in the last capstone submission, i had CNN and imagenet experiment. but this time different CNN gave a better result and imagenet is yet to match the outcome of CNN. Only new thing learnt is that simplest conv2d (3x3) with max pooling can give better result.
  * as part of new learning , I have :
    * tf eager execution, instead of graph based as in last submission
    * image generator to simplify code, augment the images and to reshape( not recommended).
    * keras way of working with transfer learning
    * numpy to image save
    * experimenting with workers and parallel processing at "fit" and "fit_generator".
    * couple of experiments were not possible if TPU and colab was not arround.
    * bokeh experiments with images.
    * automating data load operation through linux commands (wget, unzipping, creation and deletion folders, moving files, listing files)

* Iteration 5:
  * Saving models:
    * I guess below approach is good:
      * while(1):
          * model.fit()
          * model.save('./final_model.h5', include_optimizer=True)

* Iteration 6:
  * Simple LR has too many loss
  * NN also has high loss and hence low accuracy 
  * NN has limitation number of neurons with which python crashes if the initial neuron shape crosses 8K.


* iteration 7:
  * PCA + Normalization ( scaling between 0 to 1 ) gave a better result 40 % with PCA features of 1024
    * PCA training /fit took half an hour
    * Dense layer training took mere 3 min for 100 epochs

  * PCA + Normalization ( scaling between 0 to 1 ) gave even better result 45%
    * PCA training/fit took 1hr 15 min
    * dense layer took around 5-8 min for 100 epochs

  * evidentally, it is all about minimizing the loss. More number of features in PCA gives higher accuracy is the clear indication of lossess due to drastic step down from 200k to 1k or 2k neurons. So what are the losses we have in our pipeline?
    * images are saved as jpg; at that time we have loss of scaling it to 0-255 UNSIGNED INTEGER.
    * later when we do a normalization the loss is compounded.
    * there may be loss in scaling and doing pca due to sklearn capability to work on float64 (if such constraints exists)


* Iteration 8:
  * WTF !!!!, np.int for image array gave LR an accuracy of 45%
  * np.float64 for image array gave 72% accuracy
  * np.float128 for image array gave 76% for LR classification.
  * However, as anticipated Neural Network dint in herit this accuracy change.(45%)
      * I though additional layers would fine tune the 1 LR layer but it seems they have introduced loss ( value of 8 for NN against 3 in LR). Where is the loss
          * RELU : there may have been loss due to negative activation values. Just like that of PCA can have -ve value an activation value might have been negative and might have been thrown away by RELU.
            * one more reason to suspect is that changes in the network apart from additional layer is introduction of activation function relu.

        * Sigmoid : tried randomly sigmoid function. Pictorically, I can remember that the graph looks like letter "S" (slant), so I know that it handles either side of axes and hence should help if my hypothesis on relu loss.Loss definitely got reduced. Loss value looks better (0.9) when compared to both LR (3)  and relu(8)
          * training took more 80 secs, because i vaguley remember that sigmoid has exponential function in its equation
          * However, it looks promising, to run more epochs in search of reduced loss and increased accuracy.
            * Note that here training accuracy was max at 66% but testing at 42% but it evident that epochs are too low (5 epochs).

        * tanh: I know that it does not have exponential functions. Indeed there was reduction in training time (20 secs less per epoch compared to sigmoid : near 60 secs)
            * accuracy around 45% again not enough epochs to conclude

        * linear: suicidal.(as per my hypothesis) Negative values has to be controlled. Linear does NOT do that. LOSS is huge !!!
          * training time : 1 min around, accuracy = 0.3
        
        * softsign: miniature version of sigmoid. <1 min training time , promissing accuracy ( 60:50).
          * unlike expected model did not learn on increased epochs

        * selu: Surprisingly failed. 
          * 61s 1s/step - loss: 11.3066 - accuracy: 0.3044 - Recall: 0.3044 - Precision: 0.3044


    * PCA computation mandates that we need to have the entire dataset in memory and hence it is failing in my case.
        * float128 for image array consumes most of the RAM leaving no room for PCA or NN

Iteration 9: 
  * Data loading is a simplest task but I have made so many revisions
    * with view of minimizing loss, fresh read from the file is consuming 80 seconds per epoch from few seconds!!!
  * tensorflow does not support float128, max of float64 is feasible.
  * new design is to create batches and save the batches and load it when generator is called.
  * random.choices does not work; need to use random.sample instead



# To Do
* Batch normalization for both CNN and LR
* dropout between each dense layer
* is LR with PCA ( n = say 10K) equal to NN with cone structure ?
  * as in, cone structure (decreasing stacked NN layer neurons) reduces the feature vector at each layer so is this a parallel approach to PCA ?

* NN Experiments:
  * adding addtional CNN layer to make the last dimention as 14x14 or 6x6
      * 6x6 took additional time for one epoch and also took addional epochs to reach the accuracy as that of 14x14.
      
  * PCA and LR:
    * 1 million (half of flattened layer) PCA dimention
    * 50K PCA dimention
    * 10K PCA dimention
    * 5k
    * 1K
    * 512
    * 256
  * NN layers
    * 100K, 50K, 25K, 12K, 6K, 3K, 1K, 512, 256, 128, 64, 32, 16, 8
      * batch normalization
      * drop outs

* Preprocessing:
  * skull removal through clustering
  * image generator to reduce the size to 256x256


* transfer learning
* image segmentation
  * try with only window of image with tumor for accuracy
  * how is image segmentation actually done
  

In [0]:
for i in train_generator:
  print(i[1].shape)
  break

In [0]:
np.array(df_flat.label).shape

In [0]:
def imageflatdf():
  imageflatlist=[]
  label_transform = [[],np.array([0,0,1]),np.array([0,1,0]),np.array([1,0,0])]
  #for d in os.listdir( "training_data/"):
  #  for f in os.listdir("training_data/"+d+"/"):
  for file_name in os.listdir("download/mat"):
      #plt.imshow(plt.imread("training_data/1/"+f))
      #print(plt.imread("training_data/1/"+f).reshape(1,-1).shape)
      with h5py.File("download/mat/"+file_name,'r') as f:
        image_array = np.array(f['cjdata']['image'],dtype=np.float128)
        label = np.array(f['cjdata']['label'], dtype=np.int)[0][0]
        image_array = image_array/image_array.max()
        if image_array.shape[0] == 512:
          imageflatlist.append((list(image_array.reshape(-1)),label_transform[label]))
        else:
          image_array = np.pad(image_array,(512 - image_array.shape[0])//2,'constant',constant_values=0)
          imageflatlist.append((list(image_array.reshape(-1)),label_transform[label])) 
      #break
  df = pd.DataFrame(imageflatlist,columns=['image_array','label'])
  return df

df_flat = imageflatdf()
df_flat.sample(10)

In [0]:
def test(df):
    from sklearn.decomposition import PCA, IncrementalPCA
    pca = PCA(n_components=1024,whiten=True)
    #pca = IncrementalPCA(n_components=1024, batch_size=10)
    pca.fit(df.sample(1025).image_array.to_list())
    return pca

pca = test(df_flat.copy())


In [0]:
df_flat.info()

In [0]:
df_flat.groupby("label").agg("count")

In [0]:
df_flat.image_array[0].shape

#PCA + LR

In [0]:
(len(os.listdir("/content/validation_data/npz/"))//32)

In [29]:
def doLR():
  print("Start :", datetime.datetime.now())

  model = tf.keras.models.Sequential([
            tf.keras.layers.Dense(3, input_shape=[512*512], activation="softmax") 
            ])
  model.compile(loss="categorical_crossentropy"
              ,optimizer= "adam"
              ,metrics=["accuracy","Recall","Precision"])
  print(model.summary())

  batch_size=32
  model.fit_generator(myTrainGenerator(batch_size)
          ,epochs=100
          ,steps_per_epoch=len(os.listdir("training_data/npz"))
          ,validation_data=myValidateGenerator(batch_size)
          ,validation_steps=len(os.listdir("validation_data/npz/"))
          )
  
  return model.evaluate(myTestGenerator(batch_size))

doLR()
print("End :", datetime.datetime.now())

Start : 2019-10-12 23:37:20.879716
Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 3)                 786435    
Total params: 786,435
Trainable params: 786,435
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/100
68/68 [==============================] - 16s 239ms/step - loss: 4.6158 - accuracy: 0.6681 - Recall: 0.6615 - Precision: 0.6716 - val_loss: 4.0474 - val_accuracy: 0.7429 - val_Recall: 0.7429 - val_Precision: 0.7429
Epoch 2/100
68/68 [==============================] - 15s 217ms/step - loss: 4.5219 - accuracy: 0.6984 - Recall: 0.6984 - Precision: 0.6984 - val_loss: 4.1883 - val_accuracy: 0.7257 - val_Recall: 0.7257 - val_Precision: 0.7257
Epoch 3/100
68/68 [==============================] - 15s 217ms/step - loss: 4.1896 - accuracy: 0.7221 - Recall: 0.7221 - Precision: 0.7221 - val_loss:

KeyboardInterrupt: ignored

def doLR(df):
  print("Start :", datetime.datetime.now())
  #df['label'] = pd.Categorical(df['label'])
  #df['label'] = df.label.astype(np.int)

  df_test = df.sample(frac=0.2)
  df_train = df.drop(df_test.index)
  

  model = tf.keras.models.Sequential([
            tf.keras.layers.Dense(3, input_shape=[262144], activation="softmax") 
            ])
  model.compile(loss="categorical_crossentropy"
              ,optimizer= "adam"
              ,metrics=["accuracy","Recall","Precision"])
  print(model.summary())


  dataset = tf.data.Dataset.from_tensor_slices((df_train.image_array, df_train.label))
  
  for feat, targ in dataset.take(5):
    print ('Features: {}, Target: {}'.format(feat, targ))

  train_dataset = dataset.batch(50)
  #model.fit(dataset)
  model.fit_generator(train_dataset,epochs=100)
  
  dataset = tf.data.Dataset.from_tensor_slices((df_test.image_array, df_test.label))
  test_dataset = dataset.batch(50)
  return model.evaluate(test_dataset)

doLR(df_flat.copy())
print("End :", datetime.datetime.now())

def tfData(df):
  df_test = df.sample(frac=0.2)
  df_train = df.drop(df_test.index)
  train_dataset = tf.data.Dataset.from_tensor_slices((df_train.image_array, df_train.label)).batch(50)
  test_dataset = tf.data.Dataset.from_tensor_slices((df_test.image_array, df_test.label)).batch(50)
  return train_dataset, test_dataset

train_dataset, test_dataset = tfData(df_flat.copy())

def myGenerator(df,batch_size):
  for i in range(0, df.shape[0],batch_size):
    yield df.image_array.values[i:i+batch_size], df.label.values[i:i+batchsize]

In [31]:
def doNN():
  print("Start :", datetime.datetime.now())
  
  activation_function = "softsign"
  model = tf.keras.models.Sequential([
            tf.keras.layers.Dense(4098, input_shape=[512*512], activation=activation_function) 
            ,tf.keras.layers.Dropout(0.2)
            ,tf.keras.layers.Dense(2048, activation=activation_function) 
            ,tf.keras.layers.Dropout(0.2)
            ,tf.keras.layers.Dense(1024, activation=activation_function) 
            ,tf.keras.layers.Dropout(0.2)
            ,tf.keras.layers.Dense(512, activation=activation_function) 
            ,tf.keras.layers.Dropout(0.2)
            ,tf.keras.layers.Dense(128, activation=activation_function) 
            ,tf.keras.layers.Dropout(0.2)
            ,tf.keras.layers.Dense(3, activation="softmax") 
            ])
  model.compile(loss="categorical_crossentropy"
              ,optimizer= "adam"
              ,metrics=["accuracy","Recall","Precision"])
  print(model.summary())

  batch_size=32
  model.fit_generator(myTrainGenerator(batch_size)
          ,epochs=100
          ,steps_per_epoch=len(os.listdir("training_data/npz"))
          ,validation_data=myValidateGenerator(batch_size)
          ,validation_steps=len(os.listdir("validation_data/npz/"))
          )
  
  return model.evaluate(myTestGenerator(batch_size))

doNN()
print("End :", datetime.datetime.now())

Start : 2019-10-13 00:10:07.334146
Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 4098)              1074270210
_________________________________________________________________
dropout_1 (Dropout)          (None, 4098)              0         
_________________________________________________________________
dense_5 (Dense)              (None, 2048)              8394752   
_________________________________________________________________
dropout_2 (Dropout)          (None, 2048)              0         
_________________________________________________________________
dense_6 (Dense)              (None, 1024)              2098176   
_________________________________________________________________
dropout_3 (Dropout)          (None, 1024)              0         
_________________________________________________________________
dense_7 (Dense)    

KeyboardInterrupt: ignored


def doNN(df):
  #tf.reset_default_graph()
  #df['label'] = pd.Categorical(df['label'])
  #df['label'] = df.label.astype(np.int)
  print("Start :", datetime.datetime.now())
  df_test = df.sample(frac=0.2)
  df_train = df.drop(df_test.index)
  
  activation_function = "softsign"
  model = tf.keras.models.Sequential([
            tf.keras.layers.Dense(4098, input_shape=[262144], activation=activation_function) 
            ,tf.keras.layers.Dropout(0.1)
            ,tf.keras.layers.Dense(2048, activation=activation_function) 
            ,tf.keras.layers.Dropout(0.1)
            ,tf.keras.layers.Dense(1024, activation=activation_function) 
            ,tf.keras.layers.Dropout(0.1)
            ,tf.keras.layers.Dense(512, activation=activation_function) 
            ,tf.keras.layers.Dropout(0.1)
            ,tf.keras.layers.Dense(128, activation=activation_function) 
            ,tf.keras.layers.Dropout(0.1)
            ,tf.keras.layers.Dense(3, activation="softmax") 
            ])
  model.compile(loss="categorical_crossentropy"
              ,optimizer= "adam"
              ,metrics=["accuracy","Recall","Precision"])
  print(model.summary())


  #dataset = tf.data.Dataset.from_tensor_slices((df_train.image_array, df_train.label))
  
  #for feat, targ in train_dataset.take(5):
  #  print ('Features: {}, Target: {}'.format(feat, targ))

  #train_dataset = dataset.batch(50)
  #model.fit(dataset)
  model.fit_generator(train_dataset, epochs=5)
  
  #dataset = tf.data.Dataset.from_tensor_slices((df_test.image_array, df_test.label))
  #test_dataset = dataset.batch(50)
  return model.evaluate(test_dataset)

doNN(df_flat.copy())
print("End :", datetime.datetime.now())

In [0]:
class myPca:
  def doPCANN(df):
    print("Start :", datetime.datetime.now())
    #df['label'] = pd.Categorical(df['label'])
    #df['label'] = df.label.astype(np.int)
    #tf.reset_default_graph()
    #tf.Session().reset()

    from sklearn.decomposition import PCA, IncrementalPCA
    pca = PCA(n_components=1024,whiten=True)
    #pca = IncrementalPCA(n_components=1024, batch_size=10)
    pca.fit(df.sample(1025).image_array.to_list())
    scaler = MinMaxScaler()
    df["image_array_pca"] = df.image_array.apply(lambda x: scaler.fit_transform(pca.transform(x.reshape(1, -1))[0].reshape(1, -1))[0])
    print("Post PCA :", datetime.datetime.now())
    df_test = df.sample(frac=0.2)
    df_train = df.drop(df_test.index)

    model = tf.keras.models.Sequential([
              #tf.keras.layers.Dense(4098, input_shape=[1024], activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              #,tf.keras.layers.Dense(2048, activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              #,tf.keras.layers.Dense(1024, activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              #,tf.keras.layers.Dense(512, activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              #,tf.keras.layers.Dense(128, activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              tf.keras.layers.Dense(3, input_shape=[1024], activation="softmax") 
              ])
    model.compile(loss="categorical_crossentropy"
                ,optimizer= "adam"
                ,metrics=["accuracy","Recall","Precision"])
    print(model.summary())


    dataset = tf.data.Dataset.from_tensor_slices((df_train.image_array_pca, df_train.label))
    
    for feat, targ in dataset.take(5):
      print ('Features: {}, Target: {}'.format(feat, targ))

    train_dataset = dataset.batch(50)
    #model.fit(dataset)
    model.fit_generator(train_dataset,epochs=5)
    
    dataset = tf.data.Dataset.from_tensor_slices((df_test.image_array_pca, df_test.label))
    test_dataset = dataset.batch(50)
    return model.evaluate(test_dataset)

print("Test Result :",myPca.doPCANN(df_flat.copy()))
print("End :", datetime.datetime.now())

In [0]:
def doPCA(df):
  import numpy as np
  from sklearn.decomposition import PCA
  pca = PCA(n_components=2,whiten=True)

  print(np.array(df.image_array.to_list()).shape)
  pca.fit(df.image_array.to_list())
  #pca.transform(image[0].reshape((1,-1)))
  #return [pca.transform(image) for image in df.image_array.to_list() if image.shape[0] == 262144]
  df['x'] = df.image_array.apply(lambda x: pca.transform(x.reshape(1, -1))[0][0] if x.shape[0]== 262144 else 0)
  df['y'] = df.image_array.apply(lambda x: pca.transform(x.reshape(1, -1))[0][1] if x.shape[0]== 262144 else 0)

  return df

x_y = doPCA(df_flat.copy())

In [0]:
x_y.head()

In [0]:
x_y['label'] = x_y.label.astype(np.int)
x_y.info()

In [0]:
def plotp(x_y):
  plt.scatter(x_y['x'],x_y['y'],c=x_y['label'])
  plt.show()

plotp(x_y[x_y.label == 1])
plotp(x_y[x_y.label == 2])
plotp(x_y[x_y.label == 3])
plotp(x_y)

In [0]:
plt.imshow(plt.imread("training_data/1/1.jpg"))

In [0]:
temp_image = plt.imread("training_data/1/1.jpg").copy()
ma = temp_image.max()*.8
print(ma)
temp_image[temp_image > ma]= 0
plt.imshow(temp_image)

In [0]:
def doCluster(df):
  from sklearn import cluster
  k_means = cluster.KMeans(n_clusters=2, n_init=4)
  k_means.fit(df.image_array.to_list())

  return df

tt = doCluster(df_flat.copy())

##### Incremental PCA

In [0]:
def returnImageLabel(file_list):
  image_list=[]
  label_transform = [[],np.array([0,0,1]),np.array([0,1,0]),np.array([1,0,0])]
  for file_name in file_list:
    
    with h5py.File("download/mat/"+file_name,'r') as f:
          image_array = np.array(f['cjdata']['image'],dtype=np.float128)
          image_array = image_array/image_array.max()
          label = np.array(f['cjdata']['label'], dtype=np.int)[0][0]
          if image_array.shape[0] == 512:
            image_list.append((list(image_array.reshape(-1)),label_transform[label]))
          else:
            image_array = np.pad(image_array,(512 - image_array.shape[0])//2,'constant',constant_values=0)
            image_list.append((list(image_array.reshape(-1)),label_transform[label])) 
  return np.array(image_list)

def myGenerator(batch_size):
  files = os.listdir("download/mat")
  for i in itertools.cycle(range(0,len(files),batch_size)):
    yield returnImageLabel(files[i:i+batch_size])

    

In [0]:
class myPca:
  def doPCANN():
    print("Start :", datetime.datetime.now())
    #df['label'] = pd.Categorical(df['label'])
    #df['label'] = df.label.astype(np.int)
    #tf.reset_default_graph()
    #tf.Session().reset()

    from sklearn.decomposition import PCA, IncrementalPCA
    pca = PCA(n_components=1024,whiten=True)
    #pca = IncrementalPCA(n_components=1024, batch_size=10)
    pca.fit(df.sample(1025).image_array.to_list())
    scaler = MinMaxScaler()
    df["image_array_pca"] = df.image_array.apply(lambda x: scaler.fit_transform(pca.transform(x.reshape(1, -1))[0].reshape(1, -1))[0])
    print("Post PCA :", datetime.datetime.now())
    df_test = df.sample(frac=0.2)
    df_train = df.drop(df_test.index)

    model = tf.keras.models.Sequential([
              #tf.keras.layers.Dense(4098, input_shape=[1024], activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              #,tf.keras.layers.Dense(2048, activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              #,tf.keras.layers.Dense(1024, activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              #,tf.keras.layers.Dense(512, activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              #,tf.keras.layers.Dense(128, activation="relu") 
              #,tf.keras.layers.Dropout(0.5)
              tf.keras.layers.Dense(3, input_shape=[1024], activation="softmax") 
              ])
    model.compile(loss="categorical_crossentropy"
                ,optimizer= "adam"
                ,metrics=["accuracy","Recall","Precision"])
    print(model.summary())


    dataset = tf.data.Dataset.from_tensor_slices((df_train.image_array_pca, df_train.label))
    
    for feat, targ in dataset.take(5):
      print ('Features: {}, Target: {}'.format(feat, targ))

    train_dataset = dataset.batch(50)
    #model.fit(dataset)
    model.fit_generator(train_dataset,epochs=5)
    
    dataset = tf.data.Dataset.from_tensor_slices((df_test.image_array_pca, df_test.label))
    test_dataset = dataset.batch(50)
    return model.evaluate(test_dataset)

print("Test Result :",myPca.doPCANN())
print("End :", datetime.datetime.now())